### ISSUE TITLE
Developing hooks interactively

### ISSUE THREAD

**Issue by sverhagen** [pre-commit/pre-commit#589]
It appears to me that what is described as "Developing hooks interactively" is very limited in its usage, but I'm probably just misunderstanding how to do it. While I can point to a Git repository on my file system, I still need to commit in _that_ repository, and push to the remote, since `pre-commit autoupdate` or similar commands try to `git clone` and `git reset X --hard` where `X` is a SHA or a branch name or something like that. It would be more useful to point to a file location, and make ongoing changes there, test the changes iteratively, _then_ commit these changes. It is also unclear what the meaning of `sha: ...` is in this configuration. There may not really be a problem here, except that I don't understand things :wink:

**Comment by asottile** [comment]
Here's how I tend to do local development against hooks (when necessary, usually I only do it for the first iteration of the hook and then rely on other tests in the repository itself to manage the hooks).

Initialize my git repository, Let's say it's at `/home/asottile/workspace/hook-repo`

Do some development (on `master` branch -- yeah this part is not the greatest) and commit a `.pre-commit-hooks.yaml` file.

Run `git rev-parse HEAD` to find your current revision (we'll only need to do this for the first time, we'll rely on autoupdate from then on).

With this revision, temporarily add to a consuming repository the following:

```yaml
-   repo: /home/asottile/workspace/hook-repo
    sha: <<the result of git rev-parse HEAD>>
    hooks:
    -   your-hook-id
```

Continue to iterate by doing the following:

- In your `hook-repo` make whatever changes are necessary
- Run `git commit ...` (since pre-commit works by resetting to a specific revision, an actual commit / tag is necessary for `sha: ...`)
- In your consuming repo, run `pre-commit autoupdate --bleeding-edge` to update to the latest revision.  This will automatically set the `sha: ...` to the latest revision on the `master` branch (without `--bleeding-edge`, it will seek out the nearest tag).
- Then you can test with `pre-commit run --all-files` or `pre-commit run --files ...` etc.
- If you want to do a proper pull request to your `hook-repo`: once satisfied, in `hook-repo` rename the `master` branch to something appropriate for a pull request and rebase / etc. your branch.
    - You can rename a branch by running `git branch -m my-new-branch-name`

You shouldn't need to push if you're referring to the local filesystem (just commit) as a filesystem location is a valid clone source.

Let me know if any part of this is unclear or what can be added to the documentation (or if you want to PR the docs, they [live here](https://github.com//pre-commit/pre-commit.github.io) :)

**Comment by asottile** [comment]
@sverhagen any thoughts on this?

**Comment by sverhagen** [comment]
It's after midnight on a Sunday night, for the second week in a row, from all possible times, _that_ is when I chose to work on this little side project. Just to say sorry for not being very responsive, and probably not at my very best game. I didn't expect a commit in the hooks repository to be necessary, so I started doing that now. Also, I am on a branch in my hooks repository, while I now understand that you're saying that I need to be on `master`, in order to use `autoupdate`.  Even without `--bleeding-edge`, an `autoupdate` is clearly looking at `master`, which is not where my new stuff is. I'm now "solving" that by doing the following workflow:

1. Edit hooks
2. Commit hooks repository (on a story branch) (don't push)
3. `rm -rf /home/sverhagen/.pre-commit`
4. Commit test repository
5. Repeat; go to 1.
6. When all done, groom the subsequent commits in 2. into a single, decent commit, then commit hooks repository (on a story branch) (push my branch)

The `.pre-commit-config.yaml` in my test repository has `sha: <name_of_the_branch>`.


Evidently, I'm making it work. But it's not all too pretty.

Personally I feel it'd be useful for development if any steps between 1. and 4. could be eliminated. Maybe we can come up with some value for `sha` that tells pre-commit to copy my given file path, rather then `git clone`/`pull` from it (say, "dev mode"). But I don't know pre-commit well enough yet to know if that would make sense in the bigger scheme of things.

**Comment by asottile** [comment]
Would a cli interface such as `pre-commit test-repo path/to/repo hook-id` satisfy your requirements?

**Comment by mwarkentin** [comment]
@asottile Would that run the hook or just install it? 

Another option might be to allow `--bleeding-edge` to support a branch argument so that we could auto-update from our branch for testing before merging to master.

**Comment by asottile** [comment]
It could do whatever, I was imagining it to install + run the hook based on a git directory.

Adding `--branch` to autoupdate would be pretty easy (for use with `--bleeding-edge`), I can add that :)

**Comment by mwarkentin** [comment]
Great, I think that would be helpful! I also like the `test-repo` idea to install and run the hook.

**Comment by asottile** [comment]
I decided to call this `try-repo` instead as it has some nice useful characteristics outside developing hooks.

Here's the PR for that: #633

**Comment by asottile** [comment]
So the new workflow with `try-repo` now looks something like this:

- HR: hook repository
- CR: consuming repository

1. HR: on feature branch `git commit -m "..."` and/or `git commit --amend` (still need a commit for it to be clonable)
2. either:
    - CR: `pre-commit test-repo ../path/to/hook/repo`
    - CR: `pre-commit test-repo ../path/to/hook/repo specific-hook-id`

**Comment by asottile** [comment]
This has been released in v1.3.0

I'm going to call this fixed for now, I'll update the documentation to call out the new functionality :)

**Comment by mwarkentin** [comment]
@asottile Nice, thanks!

**Comment by sverhagen** [comment]
Just wanted to quickly circle back to this. I've changed employers in the meantime, and with that the immediate need have gone away, along with my access to the code base where this applied to. I just want to express my appreciation for your quick and eager follow up on my issue, that's great! Thank you!

**Comment by revolter** [comment]
I think it would still be a big improvement to allow testing a hook without the need of a commit.

Also, `try-repo` completely ignores the CR's config file (e.g. the hook's `files` parameters).

**Comment by asottile** [comment]
I agree that no-commit would be ideal.   I wonder if there's some way to abuse [`git-write-tree`](https://git-scm.com/docs/git-write-tree) / [`git-read-tree`](https://git-scm.com/docs/git-read-tree) and a custom `GIT_INDEX_FILE` to make something clonable?

Could you elaborate on ignoring the config file?  It seems to work for me:

```
$ cat ../1/.pre-commit-hooks.yaml 
-   name: test
    id: test
    language: system
    entry: echo
    files: \.py$
$ git ls-files
foo.txt
setup.py
$ pre-commit try-repo ../1 --all-files --verbose
[INFO] Initializing environment for ../1.
===============================================================================
Using config:
===============================================================================
repos:
-   repo: ../1
    sha: 98562dd751fb2b7cfe04d2171cce5198dd5adaac
    hooks:
    -   id: test
===============================================================================
[test] test..............................................................Passed
hookid: test

setup.py

```

**Comment by revolter** [comment]
If I'm not wrong, pre-commit already knows how to save and restore patches with the changes in the working tree. Could it use it also for this case?

```bash
> cat .pre-commit-config.yaml
repos:
-   repo: <redacted>
    sha: v0.0.1
    hooks:
    -   id: check-iap-test
        files: >
            (?x)(
                FileA|
                FileB
            )
    -   id: check-rubocop

> pre-commit try-repo ../Utils/my-hooks --all-files --verbose
[INFO] Initializing environment for ../Utils/my-hooks.
===============================================================================
Using config:
===============================================================================
repos:
-   repo: ../Utils/my-hooks
    sha: c08ea92571ba396265d031ca8c9ee83d7ca2c46e
    hooks:
    -   id: check-iap-test
===============================================================================
[INFO] Installing environment for ../Utils/my-hooks.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
[check-iap-test] Check IAP test..........................................
```

As you can see, the used config is missing the `files` parameter.

**Comment by asottile** [comment]
The patching/restore patch bit is a bit different.  This needs to make something clonable (since hook repositories are installed by _cloning_).

The other thing is a confusion I believe.  `try-repo` makes its _own_ `.pre-commit-config.yaml` that it uses to run hooks.  I think you're expecting them to be merged (somehow?) -- though I don't know how you'd expect that to work other than guessing that hookid is globally unique (it isn't) (there's nothing else that matches between the two). `try-repo` is designed to work even on _empty_ repositories.

**Comment by revolter** [comment]
Well, I definitely expected it to use my project's config file. Without it, the test doesn't make any sense at all, as I want for them to run for specific files.

**Comment by asottile** [comment]
Use `--files` perhaps?

**Comment by asottile** [comment]
I've devised a way to make unstaged changes clonable -- would need to make this more robust and teach `try-repo` how to do this -- but it seems to work?

### test.sh

```bash
#!/usr/bin/env bash
set -euxo pipefail

SRC="$PWD/1"
SHADOW="$PWD/2"

rm -rf "${SHADOW}"

head="$(git -C "${SRC}" rev-parse HEAD)"

: show the unstaged differences in the src repository
git -C "${SRC}" diff

: make a clone of the current repo
git clone --no-checkout "${SRC}" ${SHADOW}

: make a branch so we will have a reachable revision
git -C "${SHADOW}" checkout "${head}" -b _pc_unstaged

: Add the current files in this repo to the index in the other
GIT_INDEX_FILE="${SHADOW}/.git/index" git -C "${SRC}" add --all :/

: Commit to generate a commit
git -C "${SHADOW}" commit --allow-empty -m 'add unstaged changes'

: 'success!'
git -C "${SHADOW}" show
```

### output

```console
$ ./test.sh 
+ SRC=/tmp/wat/1
+ SHADOW=/tmp/wat/2
+ rm -rf /tmp/wat/2
++ git -C /tmp/wat/1 rev-parse HEAD
+ head=98562dd751fb2b7cfe04d2171cce5198dd5adaac
+ : show the unstaged differences in the src repository
+ git -C /tmp/wat/1 diff
diff --git a/.pre-commit-hooks.yaml b/.pre-commit-hooks.yaml
index 5c1a852..f917278 100644
--- a/.pre-commit-hooks.yaml
+++ b/.pre-commit-hooks.yaml
@@ -1,5 +1,5 @@
 -   name: test
     id: test
     language: system
-    entry: echo
+    entry: echo ohai
     files: \.py$
+ : make a clone of the current repo
+ git clone --no-checkout /tmp/wat/1 /tmp/wat/2
Cloning into '/tmp/wat/2'...
done.
+ : make a branch so we will have a reachable revision
+ git -C /tmp/wat/2 checkout 98562dd751fb2b7cfe04d2171cce5198dd5adaac -b _pc_unstaged
Switched to a new branch '_pc_unstaged'
+ : Add the current files in this repo to the index in the other
+ GIT_INDEX_FILE=/tmp/wat/2/.git/index
+ git -C /tmp/wat/1 add --all :/
+ : Commit to generate a commit
+ git -C /tmp/wat/2 commit --allow-empty -m 'add unstaged changes'
[_pc_unstaged b727c9a] add unstaged changes
 1 file changed, 1 insertion(+), 1 deletion(-)
+ : 'success!'
+ git -C /tmp/wat/2 show
commit b727c9a22a9dd5fe880ec4c0ac729ade9b5fac9b (HEAD -> _pc_unstaged)
Author: Anthony Sottile <asottile@umich.edu>
Date:   Tue Feb 13 10:16:23 2018 -0800

    add unstaged changes

diff --git a/.pre-commit-hooks.yaml b/.pre-commit-hooks.yaml
index 5c1a852..f917278 100644
--- a/.pre-commit-hooks.yaml
+++ b/.pre-commit-hooks.yaml
@@ -1,5 +1,5 @@
 -   name: test
     id: test
     language: system
-    entry: echo
+    entry: echo ohai
     files: \.py$
```

**Comment by revolter** [comment]
Indeed is a workable solution, but, do you agree that a regular user would expect for the project's config file to be loaded? There are more cases in which you want to test a hook in the context of your project, instead of in an abstract way.

**Comment by asottile** [comment]
I agree they could, but they'd just be confused :) -- this is part of the reason that `try-repo` prints out exactly what config it is about to run.

I've opened another issue for handling unstaged changes: https://github.com/pre-commit/pre-commit/issues/703

**Comment by revolter** [comment]
What about an option argument that allows this? 😃

**Comment by asottile** [comment]
For some value of _this_, yeah an option would probably be fine

**Comment by asottile** [comment]
I adjusted my little script above to also use `GIT_OBJECT_DIRECTORY` which seems to be necessary on newer versions of `git` than whatever version I used at the time.

#904 is a PR which implements this

**Comment by asottile** [comment]
This has been released as part of [v1.14.0](https://github.com/pre-commit/pre-commit/releases/tag/v1.14.0) 🎉  -- thanks again!
